core/services/workflows/engine.go

```diff
@@ -669,10 +695,18 @@ func (e *Engine) workerForStepRequest(ctx context.Context, msg stepRequest) {
 	// receiving loop may not pick up any messages we emit.
 	// Note: When full persistence support is added, any hanging steps
 	// like this one will get picked up again and will be reprocessed.
+	e.stepUpdatesChMapMu.RLock()
+	stepUpdateCh, ok := e.stepUpdatesChMap[stepState.ExecutionID]
+	if !ok {
+		e.stepUpdatesChMapMu.RUnlock()
+		l.Errorf("step update channel not found for execution %s, dropping step update", stepState.ExecutionID)
+		return
+	}
+	e.stepUpdatesChMapMu.RUnlock()
 	select {
 	case <-ctx.Done():
 		l.Errorf("context canceled before step update could be issued; error %v", err)
-	case e.stepUpdateCh <- *stepState:
+	case stepUpdateCh.ch <- *stepState:
```

Comment:
I think you technically need to do a locked send here, to avoid a potential race between 699 and 709 where the channel is closed. This shouldn't be possible but to be safe we should add it.